cpp/include/cudf/detail/unique_hash_join.cuh

```diff
@@ -0,0 +1,139 @@
+    cudf::experimental::row::equality::device_row_comparator<true, cudf::nullate::DYNAMIC>>;
+  ///< Row equality type for flat columns
+  using flat_row_equal = cudf::experimental::row::equality::strong_index_comparator_adapter<
+    cudf::experimental::row::equality::device_row_comparator<false, cudf::nullate::DYNAMIC>>;
+
+  ///< Device row equal type
+  using d_equal_type =
+    std::conditional_t<HasNested == has_nested::YES, nested_row_equal, flat_row_equal>;
+  using probing_scheme_type = cuco::experimental::
+    double_hashing<DEFAULT_JOIN_CG_SIZE, thrust::identity<hash_value_type>, Hasher>;
+  using cuco_storge_type = cuco::experimental::storage<1>;
+
+  using hash_table_type = cuco::experimental::static_set<
+    cuco::pair<hash_value_type, cudf::experimental::row::lhs_index_type>,
+    cuco::experimental::extent<std::size_t>,
+    cuda::thread_scope_device,
+    comparator_adapter<d_equal_type>,
+    probing_scheme_type,
+    cudf::detail::cuco_allocator,
+    cuco_storge_type>;
+
+  bool _is_empty;   ///< true if `_hash_table` is empty
+  bool _has_nulls;  ///< true if nulls are present in either build table or any probe table
+  cudf::null_equality _nulls_equal;  ///< whether to consider nulls as equal
+  cudf::table_view _build;           ///< input table to build the hash map
+  cudf::table_view _probe;           ///< input table to build the hash map
+  std::shared_ptr<cudf::experimental::row::equality::preprocessed_table>
+    _preprocessed_build;  ///< input table preprocssed for row operators
+  std::shared_ptr<cudf::experimental::row::equality::preprocessed_table>
+    _preprocessed_probe;        ///< input table preprocssed for row operators
+  hash_table_type _hash_table;  ///< hash table built on `_build`
+
+ public:
+  unique_hash_join()                                   = delete;
+  ~unique_hash_join()                                  = default;
+  unique_hash_join(unique_hash_join const&)            = delete;
+  unique_hash_join(unique_hash_join&&)                 = delete;
+  unique_hash_join& operator=(unique_hash_join const&) = delete;
+  unique_hash_join& operator=(unique_hash_join&&)      = delete;
+
+  /// Row equality type for nested columns
+  /**
+   * @brief Constructor that internally builds the hash table based on the given `build` table.
+   *
+   * @throw cudf::logic_error if the number of columns in `build` table is 0.
+   *
+   * @param build The build table, from which the hash table is built
+   * @param probe The probe table
+   * @param has_nulls Flag to indicate if the there exists any nulls in the `build` table or
+   *        any `probe` table that will be used later for join.
+   * @param compare_nulls Controls whether null join-key values should match or not.
+   * @param stream CUDA stream used for device memory operations and kernel launches.
+   */
+  unique_hash_join(cudf::table_view const& build,
+                   cudf::table_view const& probe,
```

Comment:
Curious why the probe table is passed to the constructor rather than passed to the join method?  Requiring it in the constuctor seems to preclude a scenario where we have a single build table and want to compute a unique hash table once then reuse it as we join across the multiple probe tables.  That can be done today via hash_join.

Reply:
We are using the new `cuco::static_set` for `unique_hash_join` and it requires the key equal type to know at build time and the key equal type is determined by both probe and build tables. 

`hash_join` doesn't require this since the legay cuco hash table can take a custom key equal during probing. This feature hasn't been added to the new `cuco::static_set` yet.